File: sdks/typescript/src/session.ts
import { BudgetExhausted, LoopDetected } from "./errors.js";
import { Ledger, CostEvent } from "./ledger.js";
import { CircuitBreaker } from "./circuit_breaker.js";
import { calculateCost } from "./pricing.js";

export interface Report {
  session_id: string;
  budget: number;
  total_spent: number;
  remaining: number;
  breakdown: Record<string, unknown>;
  duration_seconds: number | null;
  terminated_by: "budget_exhausted" | "loop_detected" | null;
  event_count: number;
}

/** Minimal shape we need from an OpenAI chat completion response. */
export interface OpenAICompletionLike {
  model: string;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
  } | null;
}

/** Minimal shape we need from an Anthropic message response. */
export interface AnthropicMessageLike {
  model: string;
  usage: {
    input_tokens: number;
    output_tokens: number;
  };
}

export interface SessionOptions {
  id?: string | undefined;
  onSoftLimit?: ((report: Report) => void) | undefined;
  onHardLimit?: ((report: Report) => void) | undefined;
  onLoopDetected?: ((report: Report) => void) | undefined;
}

let _counter = 0;

function generateId(): string {
  return `sess_${Date.now()}_${++_counter}`;
}

/**
 * BudgetSession tracks costs for a single agent run.
 * Create via `Budget.newSession()`.
 */
export class BudgetSession {
  private readonly _id: string;
  private readonly _ledger: Ledger;
  private readonly _breaker: CircuitBreaker;
  private readonly _onSoftLimit?: ((r: Report) => void) | undefined;
  private readonly _onHardLimit?: ((r: Report) => void) | undefined;
  private readonly _onLoopDetected?: ((r: Report) => void) | undefined;
  private readonly _startTime: number;
  private _endTime: number | null = null;
  private _terminatedBy: Report["terminated_by"] = null;

  constructor(ledger: Ledger, breaker: CircuitBreaker, opts: SessionOptions = {}) {
    this._id = opts.id ?? generateId();
    this._ledger = ledger;
    this._breaker = breaker;
    this._onSoftLimit = opts.onSoftLimit;
    this._onHardLimit = opts.onHardLimit;
    this._onLoopDetected = opts.onLoopDetected;
    this._startTime = Date.now();
  }

  get id(): string { return this._id; }
  get spent(): number { return this._ledger.spent; }
  get remaining(): number { return this._ledger.remaining; }

  /** Returns true if the given cost would exceed the remaining budget. Does not record. */
  wouldExceed(cost: number): boolean {
    return this._ledger.wouldExceed(cost);
  }

  /**
   * Record the cost of an OpenAI chat completion response.
   * Throws BudgetExhausted or LoopDetected if limits are reached.
   */
  wrapOpenAI<T extends OpenAICompletionLike>(response: T): T {
    if (!response.usage) return response;
    const { prompt_tokens, completion_tokens } = response.usage;
    this._recordLLM(response.model, prompt_tokens, completion_tokens);
    return response;
  }

  /**
   * Record the cost of an Anthropic message response.
   * Throws BudgetExhausted or LoopDetected if limits are reached.
   */
  wrapAnthropic<T extends AnthropicMessageLike>(response: T): T {
    this._recordLLM(
      response.model,
      response.usage.input_tokens,
      response.usage.output_tokens
    );
    return response;
  }

  /**
   * Record the cost of any LLM call given the model name and token counts directly.
   * Throws BudgetExhausted or LoopDetected if limits are reached.
   */
  wrapUsage(model: string, inputTokens: number, outputTokens: number): void {
    this._recordLLM(model, inputTokens, outputTokens);
  }

  /**
   * Track a fixed-cost action (tool call, API call, etc.).
   * Returns the result unchanged for easy chaining.
   * Throws BudgetExhausted or LoopDetected if limits are reached.
   */
  track<T>(result: T, cost: number, toolName?: string, metadata?: Record<string, unknown>): T {
    const event: CostEvent = {
      cost,
      type: "tool",
      timestamp: Date.now(),
      toolName,
      metadata,
    };
    this._record(event, toolName);
    return result;
  }

  /**
   * Create a child session with a sub-budget.
   * When the child is done, roll up its spend with:
   *   parent.track(null, child.spent, "child:" + child.id)
   */
  childSession(maxSpend: number, opts?: SessionOptions): BudgetSession {
    const cap = Math.min(maxSpend, this.remaining);
    const childLedger = new Ledger(cap);
    const childBreaker = new CircuitBreaker(
      this._breaker["softLimitFraction"],
      this._breaker["loopMaxCalls"],
      this._breaker["loopWindowMs"] / 1000
    );
    return new BudgetSession(childLedger, childBreaker, opts ?? {});
  }

  /** Mark the session as ended. */
  close(): void {
    if (this._endTime === null) {
      this._endTime = Date.now();
    }
  }

  /** Structured cost report. */
  report(): Report {
    const endMs = this._endTime ?? Date.now();
    const durationSecs = Math.round((endMs - this._startTime) / 10) / 100;

    return {
      session_id: this._id,
      budget: this._ledger.budget,
      total_spent: round6(this._ledger.spent),
      remaining: round6(this._ledger.remaining),
      breakdown: this._ledger.breakdown(),
      duration_seconds: durationSecs,
      terminated_by: this._terminatedBy,
      event_count: this._ledger.events().length,
    };
  }

  private _recordLLM(model: string, inputTokens: number, outputTokens: number): void {
    const cost = calculateCost(model, inputTokens, outputTokens);
    if (cost === undefined) return; // unknown model — skip silently

    const event: CostEvent = {
      cost,
      type: "llm",
      timestamp: Date.now(),
      model,
      inputTokens,
      outputTokens,
    };
    this._record(event, model);
  }

  private _record(event: CostEvent, loopKey?: string): void {
    try {
      this._ledger.record(event);
    } catch (err) {
      if (err instanceof BudgetExhausted) {
        this._terminatedBy = "budget_exhausted";
        this._endTime ??= Date.now();
        this._onHardLimit?.(this.report());
      }
      throw err;
    }

    // Soft limit check (fires once)
    const warning = this._breaker.checkBudget(this._ledger.spent, this._ledger.budget);
    if (warning) {
      this._onSoftLimit?.(this.report());
    }

    // Loop detection
    if (loopKey && this._breaker.checkLoop(loopKey)) {
      this._terminatedBy = "loop_detected";
      this._endTime ??= Date.now();
      this._onLoopDetected?.(this.report());
      throw new LoopDetected(loopKey);
    }
  }
}

function round6(v: number): number {
  return Math.round(v * 1e6) / 1e6;
}

File: sdks/go/session.go
package agentbudget

import (
	"fmt"
	"math"
	"sync"
	"time"
)

// Report is the structured cost report returned by Session.Report().
type Report struct {
	SessionID     string         `json:"session_id"`
	Budget        float64        `json:"budget"`
	TotalSpent    float64        `json:"total_spent"`
	Remaining     float64        `json:"remaining"`
	Breakdown     map[string]any `json:"breakdown"`
	DurationSecs  *float64       `json:"duration_seconds"`
	TerminatedBy  string         `json:"terminated_by"` // "" | "budget_exhausted" | "loop_detected"
	EventCount    int            `json:"event_count"`
}

// Session tracks costs for a single agent run. Create via Budget.NewSession().
// It is safe to use from multiple goroutines.
type Session struct {
	id             string
	ledger         *Ledger
	breaker        *CircuitBreaker
	onSoftLimit    func(Report)
	onHardLimit    func(Report)
	onLoopDetected func(Report)

	mu           sync.Mutex
	startTime    time.Time
	endTime      time.Time
	terminatedBy string
}

func newSession(
	id string,
	ledger *Ledger,
	breaker *CircuitBreaker,
	onSoftLimit, onHardLimit, onLoopDetected func(Report),
) *Session {
	return &Session{
		id:             id,
		ledger:         ledger,
		breaker:        breaker,
		onSoftLimit:    onSoftLimit,
		onHardLimit:    onHardLimit,
		onLoopDetected: onLoopDetected,
		startTime:      time.Now(),
	}
}

// ID returns the session identifier.
func (s *Session) ID() string { return s.id }

// Spent returns total dollars spent so far.
func (s *Session) Spent() float64 { return s.ledger.Spent() }

// Remaining returns dollars left in the budget.
func (s *Session) Remaining() float64 { return s.ledger.Remaining() }

// WouldExceed returns true if the given cost would exceed the remaining budget.
// Does not record anything.
func (s *Session) WouldExceed(cost float64) bool { return s.ledger.WouldExceed(cost) }

// WrapUsage records the cost of a completed LLM call given its model and token counts.
// Returns *BudgetExhausted if the budget is exceeded, *LoopDetected if a loop is found.
func (s *Session) WrapUsage(model string, inputTokens, outputTokens int64) error {
	cost, ok := CalculateCost(model, inputTokens, outputTokens)
	if !ok {
		// Unknown model — skip silently (same as Python behaviour)
		return nil
	}

	event := CostEvent{
		Cost:         cost,
		Type:         CostTypeLLM,
		Timestamp:    time.Now(),
		Model:        model,
		InputTokens:  inputTokens,
		OutputTokens: outputTokens,
	}
	if err := s.ledger.Record(event); err != nil {
		s.setTerminated("budget_exhausted")
		if s.onHardLimit != nil {
			s.onHardLimit(s.Report())
		}
		return err
	}
	return s.checkAfterRecord(model)
}

// Track manually records a fixed-cost action (tool call, API call, etc.).
// Returns *BudgetExhausted or *LoopDetected if limits are reached.
func (s *Session) Track(cost float64, toolName string, metadata ...map[string]any) error {
	var meta map[string]any
	if len(metadata) > 0 {
		meta = metadata[0]
	}
	event := CostEvent{
		Cost:      cost,
		Type:      CostTypeTool,
		Timestamp: time.Now(),
		ToolName:  toolName,
		Metadata:  meta,
	}
	if err := s.ledger.Record(event); err != nil {
		s.setTerminated("budget_exhausted")
		if s.onHardLimit != nil {
			s.onHardLimit(s.Report())
		}
		return err
	}
	return s.checkAfterRecord(toolName)
}

// ChildSession creates a sub-session with its own budget cap.
// When the child is closed via Close(), its total spend is NOT automatically
// charged to the parent — call parent.Track(child.Spent(), "child:"+child.ID())
// after closing if you want rollup.
func (s *Session) ChildSession(maxSpend float64, opts ...SessionOption) *Session {
	cap := math.Min(maxSpend, s.Remaining())
	childLedger := newLedger(cap)
	childBreaker := newCircuitBreaker(
		s.breaker.softLimitFraction,
		s.breaker.loopMaxCalls,
		s.breaker.loopWindowSeconds,
	)
	childID := fmt.Sprintf("sess_%s_child", s.id)
	cfg := &sessionConfig{}
	for _, o := range opts {
		o(cfg)
	}
	if cfg.id != "" {
		childID = cfg.id
	}
	return newSession(childID, childLedger, childBreaker, nil, nil, nil)
}

// Close marks the session as ended. Safe to call multiple times.
func (s *Session) Close() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.endTime.IsZero() {
		s.endTime = time.Now()
	}
}

// Report returns a structured cost summary for this session.
func (s *Session) Report() Report {
	s.mu.Lock()
	end := s.endTime
	terminated := s.terminatedBy
	s.mu.Unlock()

	var dur *float64
	if !s.startTime.IsZero() {
		ref := end
		if ref.IsZero() {
			ref = time.Now()
		}
		d := ref.Sub(s.startTime).Seconds()
		d = math.Round(d*100) / 100
		dur = &d
	}

	spent := s.ledger.Spent()
	remaining := s.ledger.Remaining()
	events := s.ledger.Events()

	return Report{
		SessionID:    s.id,
		Budget:       s.ledger.Budget(),
		TotalSpent:   round6(spent),
		Remaining:    round6(remaining),
		Breakdown:    s.ledger.Breakdown(),
		DurationSecs: dur,
		TerminatedBy: terminated,
		EventCount:   len(events),
	}
}

// checkAfterRecord runs soft-limit and loop-detection checks after a successful record.
func (s *Session) checkAfterRecord(key string) error {
	spent := s.ledger.Spent()
	budget := s.ledger.Budget()

	if msg := s.breaker.CheckBudget(spent, budget); msg != "" {
		if s.onSoftLimit != nil {
			s.onSoftLimit(s.Report())
		}
	}

	if key != "" && s.breaker.CheckLoop(key) {
		s.setTerminated("loop_detected")
		if s.onLoopDetected != nil {
			s.onLoopDetected(s.Report())
		}
		return &LoopDetected{Key: key}
	}
	return nil
}

func (s *Session) setTerminated(reason string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.terminatedBy == "" {
		s.terminatedBy = reason
	}
	if s.endTime.IsZero() {
		s.endTime = time.Now()
	}
}
